### DISCUSSION TITLE
Issue in "Object Pose Estimation" Tutorial?

### DISCUSSION THREAD

**Discussion by coult099** [mitsuba-renderer/mitsuba3#1853] [Q&A]
Hello,

I am trying to replicate the [Obejct pose estimation tutorial](https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/object_pose_estimation.html) using the `llvm_ad_rgb` variant in Mitsuba 3.5.0, and `prb_projective` integrator. As such, I have copied the code almost exactly, but my loss _increases_ over time, and the optimized object gets into a state that is worse than the initial state as the optimization goes on. Has anyone else experienced this issue?

My initial, final, and reference images, along with my loss:
<table>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/a7df0cea-109e-4627-a8d1-925a9c25d221" width="200px"><br>
      <b>L2 Loss</b>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/aba760a4-5f83-4e77-9ad1-ae1f3d3cb3dd" width="200px"><br>
      <b>Initial</b>
    </td>
  </tr>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/6fd0165a-a03e-40af-a69e-61302de48df3" width="200px"><br>
      <b>Optimized</b>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/53baefe8-5855-4a5f-ad21-564586cd14c6" width="200px"><br>
      <b>Reference</b>
    </td>
  </tr>
</table>

My code:

```python
import drjit as dr
import mitsuba as mi
from datetime import date
from matplotlib import pyplot as plt
import os 

# Set Mitsuba variant & other parameters
mi_var = 'llvm_ad_rgb'
spp_ref = 2**8
spp_opt = 2**8
save_dir = "test_images"
integrator = {
    'type': 'prb_projective',
}
mi.set_variant(mi_var)
if os.path.isdir(save_dir) == False:
    os.makedirs(save_dir)

scene = mi.load_dict({
    'type': 'scene',
    'integrator': integrator,
    'sensor':  {
        'type': 'perspective',
        'to_world': mi.ScalarTransform4f().look_at(
                        origin=(0, 0, 2),
                        target=(0, 0, 0),
                        up=(0, 1, 0)
                    ),
        'fov': 60,
        'film': {
            'type': 'hdrfilm',
            'width': 64,
            'height': 64,
            'rfilter': { 'type': 'gaussian' },
            'sample_border': True
        },
    },
    'wall': {
        'type': 'obj',
        'filename': './scenes/meshes/rectangle.obj',
        'to_world': mi.ScalarTransform4f().translate([0, 0, -2]).scale(2.0),
        'face_normals': True,
        'bsdf': {
            'type': 'diffuse',
            'reflectance': { 'type': 'rgb', 'value': (0.5, 0.5, 0.5) },
        }
    },
    'bunny': {
        'type': 'ply',
        'filename': './scenes/meshes/bunny.ply',
        'to_world': mi.ScalarTransform4f().scale(6.5),
        'bsdf': {
            'type': 'diffuse',
            'reflectance': { 'type': 'rgb', 'value': (0.3, 0.3, 0.75) },
        },
    },
    'light': {
        'type': 'obj',
        'filename': './scenes/meshes/sphere.obj',
        'emitter': {
            'type': 'area',
            'radiance': {'type': 'rgb', 'value': [1e3, 1e3, 1e3]}
        },
        'to_world': mi.ScalarTransform4f().translate([2.5, 2.5, 7.0]).scale(0.25)
    }
})


# Render reference image & save
img_ref = mi.render(scene, seed=0, spp=spp_ref)
ref_name = save_dir + "/pose_original_" + mi_var + "_spp" + str(spp_ref) + "_" + str(date.today()) + ".png"
mi.util.write_bitmap(ref_name, img_ref)

# Get scene parameters, and save initial vertex positions in a Point3f array
params = mi.traverse(scene)
initial_vertex_positions = dr.unravel(mi.Point3f, params['bunny.vertex_positions'])

# Initialize Adam optimizer, set learning rate, and generate angle & translation parameters to optimize
opt = mi.ad.Adam(lr=0.025)
opt['angle'] = mi.Float(0.25)
opt['trans'] = mi.Point2f(0.1, -0.25)

# Function to relate angle & trans to actual vertex positions of bunny
def apply_transformation(params, opt):
    opt['trans'] = dr.clip(opt['trans'], -0.5, 0.5) # clamp to correct range
    opt['angle'] = dr.clip(opt['angle'], -0.5, 0.5)

    trafo = mi.Transform4f.translate([opt['trans'].x, opt['trans'].y, 0.0]).rotate([0, 1, 0], opt['angle'] * 100.0)

    params['bunny.vertex_positions'] = dr.ravel(trafo @ initial_vertex_positions)
    params.update()

# Apply initial transformation 
apply_transformation(params, opt)
img_init = mi.render(scene, seed=0, spp=spp_ref)
init_name = save_dir + "/pose_init_" + mi_var + "_spp" + str(spp_ref) + "_" + str(date.today()) + ".png"
mi.util.write_bitmap(init_name, img_init)

# Set optimization parameters
iteration_count = 100
loss_hist = []
# Run optimization 
for it in range(iteration_count):
    # Apply the mesh transformation
    apply_transformation(params, opt)

    # Perform a differentiable rendering
    img = mi.render(scene, params, seed=it, spp=spp_opt)

    # Evaluate the objective function
    loss = dr.mean(dr.sqr(img - img_ref))
    dr.backward(loss)
    opt.step()
    loss_hist.append(loss[0])

    print(f"Iteration {it:02d}: error={loss[0]:6f}, trans=[{opt['trans'].x[0]:.4f}, {opt['trans'].y[0]:.4f}], angle={opt['angle'][0]:.4f}", end='\r')


# render & save optimized image
opt_img = mi.render(scene, spp=spp_ref)
opt_name = save_dir + "/pose_final_" + mi_var + "_spp" + str(spp_ref) + "_" + str(date.today()) + ".png"
mi.util.write_bitmap(opt_name, opt_img)

plt.figure()
plt.plot(loss_hist)
fig_name = save_dir + "/pose_loss_" + mi_var + "_spp" + str(spp_ref) + "_" + str(date.today()) + ".png"
plt.savefig(fig_name)

```

**Comment by njroussel** [comment]
Hi @coult099

I've made minor changes to your script to run it with the most recent version of mitsuba (v3.8.1) and have not been able to reproduce the issue. 

If you think there is an issue with the gradients, I'd suggest you try to compare the forward derivatives as is done in this tutorial: https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/projective_sampling_integrators.html#Projective-sampling-integrator. I've done that, and they look close enough to the ground truth (finite differences).
